apps/vscode/src/providers/background.ts

```diff
@@ -162,6 +163,10 @@ async function setEditorHighlightDecorations(
       blockRanges.push(vscRange(block.range));
     }
 
+    // expose cell language for use in keybindings, etc
+    const language = mainLanguage(tokens);
+    vscode.commands.executeCommand('setContext', 'quartoLangId', language?.ids[0]);
+
```

Comment:
I do not feel it is appropriate to set the context here

Currently this function is quite pure. It is _just_ about setting highlight decorations and is very easy to understand.

If we take a look at _why_ it works to put this here, its because this is called from

- `triggerUpdateAllEditorsDecorations()`, which is called during the `onDidChangeVisibleTextEditors()` hook
- `triggerUpdateActiveEditorDecorations()`, which is called during the `onDidOpenTextDocument()` hook

That hook registration is part of `activateBackgroundHighlighter()`, which is called in `activateCommon()`.

---

I think what we really want is:
- In `activateCommon()`,
- Add a new `activateLanguageContextSetter()`
- Which registers `onDidOpenTextDocument()` and `onDidChangeVisibleTextEditors()` hooks of its own, and those hooks set the `quartoMainLanguageId` of the main language for the document.

Yes you have to reparse the document, but I think that is worth it over muddying the waters about what `setEditorHighlightDecorations()` is about, plus it gives you finer control over when the context is updated.

Reply:
This is a great idea! Implemented in c6f28f8673d712ba7d430fd6ee48e2e38fb39970 and working really well in https://github.com/posit-dev/positron/pull/5451.